documentcloud/documents/models.py

```diff
@@ -379,6 +380,59 @@ def get_all_page_text(self):
             )
             return {"pages": [], "updated": None}
 
+    def set_page_text(self, page_text_infos):
+        # get the json text
+        json_text = self.get_all_page_text()
+
+        # set the individual text pages
+        timestamp = int(round(time.time() * 1000))
+        json_text["updated"] = timestamp
+        if len(json_text["pages"]) < self.page_count:
+            json_text["pages"].extend(
+                [{} for _ in range(self.page_count - len(json_text["pages"]))]
+            )
+        file_names = []
+        file_contents = []
+        for page_text_info in page_text_infos:
+            page = page_text_info["page_number"]
+            text = page_text_info["text"]
+            ocr = page_text_info.get("ocr")
+            file_names.append(path.page_text_path(self.pk, self.slug, page))
+            file_contents.append(text.encode("utf8"))
+            # overwrite the text in the JSON format
+            json_text["pages"][page] = {
+                "page": page,
+                "contents": text,
+                "ocr": ocr,
+                "updated": timestamp,
+            }
+            if page_text_info.get("positions"):
+                file_names.append(
+                    path.page_text_position_path(self.pk, self.slug, page)
+                )
+                positions = [
+                    {**p.pop("metadata", {}), **p} for p in page_text_info["positions"]
+                ]
+                file_contents.append(json.dumps(positions).encode("utf-8"))
+
+        # set the full text
+        concatenated_text = b"\n\n".join(
+            [p["contents"].encode("utf-8") for p in json_text["pages"]]
+        )
+        file_names.append(path.text_path(self.pk, self.slug))
+        file_contents.append(concatenated_text)
+
+        # set the json text
+        file_names.append(path.json_text_path(self.pk, self.slug))
+        file_contents.append(json.dumps(json_text).encode("utf-8"))
+
+        # upload the text to S3
+        logger.info("[SET PAGE TEXT] upload %d", self.pk)
+        # reverse the lists to upload the larger files first
+        storage.async_upload(file_names[::-1], file_contents[::-1], access=self.access)
+
+        return json_text
```

Comment:
If we're returning before `async_upload` completes, how does the user (or anyone) know if the set text operation failed?

Reply:
So this line here does wait for `async_upload` to complete before returning, but the API call does return before this is complete (can go over in more detail live) - the short answer is there is not a good way to know at the moment.